lib/active_model/serializer/reflection.rb

```diff
@@ -42,13 +42,13 @@ class Serializer
       def build_association(subject, parent_serializer_options)
         association_value = subject.send(name)
         reflection_options = options.dup
-        serializer_class = ActiveModel::Serializer.serializer_for(association_value, reflection_options)
+        serializer_class = subject.class.serializer_for(association_value, reflection_options)
 
         if serializer_class
           begin
             serializer = serializer_class.new(
               association_value,
-              serializer_options(parent_serializer_options, reflection_options)
+              serializer_options(parent_serializer_options, reflection_options).merge(_parent_serializer: subject.class)
```

Comment:
True, I wanted to keep this PR as small as possible just to show the idea first. I could do that inside `serializer_options` (it would require an extra `subject` argument though).